Update Procedures related to Optional Features

Wind4Greg opened the pull request:
This PR addresses issue https://github.com/w3c/vc-di-bbs/issues/180. In particular it takes into account the updated APIs/procedures in the drafts:

1. [BBS Blind Signatures](https://datatracker.ietf.org/doc/html/draft-kalos-bbs-blind-signatures)
2. [BBS per Verifier Lin](https://datatracker.ietf.org/doc/html/draft-vasilis-bbs-per-verifier-linkability)

Sections [3.4.5 Base Proof Serialization (bbs-2023)](https://w3c.github.io/vc-di-bbs/#base-proof-serialization-bbs-2023), [3.4.6 Add Derived Proof (bbs-2023)](https://w3c.github.io/vc-di-bbs/#add-derived-proof-bbs-2023), [3.4.7 Verify Derived Proof (bbs-2023)](https://w3c.github.io/vc-di-bbs/#verify-derived-proof-bbs-2023) and some of their sub-procedures have been updated to use the the new APIs from these drafts. Importantly the two different flavors of *pseudonyms* now have a unified proof verification procedure which has resulted in simplified processing.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/Wind4Greg/vc-di-bbs/pull/182.html" title="Last updated on Aug 28, 2024, 2:55 PM UTC (b6e2835)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-di-bbs/182/cacfa45...Wind4Greg:b6e2835.html" title="Last updated on Aug 28, 2024, 2:55 PM UTC (b6e2835)">Diff</a>

TallTed commented on index.html:
```diff
@@ -923,6 +923,11 @@ <h4>createDisclosureData</h4>
             </li>
+          <li>
+If |featureOption| equals `"anonymous_holder_binding"`,
+`"pseudonym_issuer_pid"`, or `"pseudonym_hidden_pid"` set the |L| parameter
```
```suggestion
`"pseudonym_issuer_pid"`, or `"pseudonym_hidden_pid"`, set the |L| parameter
```

TallTed commented on index.html:
```diff
@@ -1027,7 +1032,7 @@ <h4>serializeDerivedProofValue</h4>
 selective indexes (|selectiveIndexes|), and a BBS presentation header
 (|presentationHeader|), the |featureOption| indicator, and possibly a
-|pseudonym| value depending on the |featureOption| value.
+|pseudonym| and |L| value depending on the |featureOption| value.
```
Might `|pseudonym|` be calculated without `|L|`? Or vice versa? Or are they both always present when one is?
```suggestion
|pseudonym| and/or |L| value depending on the |featureOption| value.
```

TallTed commented on index.html:
```diff
@@ -1063,41 +1068,26 @@ <h4>serializeDerivedProofValue</h4>
-Initialize |components| to an array with elements containing the values of
 |bbsProof|, |compressedLabelMap|, |mandatoryIndexes|, |selectiveIndexes|,
-|presentationHeader|, and |pseudonym|.
+|presentationHeader| and |L|.
```
```suggestion
|presentationHeader|, and |L|.
```

dlongley commented on index.html:
```diff
@@ -878,7 +878,7 @@ <h4>createDisclosureData</h4>
                 <li>
 If |featureOption| equals `"anonymous_holder_binding"`,
-set `bbsProof` to the value computed by the `ProofGen` procedure from
+set `bbsProof` to the value computed by the `BlindProofGen` procedure from
```
There's an inconsistency here between referring to the function name from the appropriate BBS spec or referring to the title of the procedure section in that spec. Here we use the function name (`BlindProofGen`) and below we use the section title `Calculate Pseudonym` and `Hidden PID Proof Generation with Pseudonym`. We should pick one of these and me consistent.

Wind4Greg replied:
@dlongley yes, when the draft IETF specs didn't have a function name (pseudonym signing) I used section names. Then I dropped the parenthesis after function names, i.e., `BlindProofGen()` is actually a function that is defined. Ideas/preferences for consistency?

dlongley replied:
I think I'd prefer the function name because it's shorter and more specific for implementers.

dlongley replied:
Bumping on this -- it can be handled in another PR if necessary, but we should get consistency at some point.

TallTed replied:
Consistency is good. Clarity is better.

Is `BlindProofGen` a function or a procedure?
Is `Calculate Pseudonym` a procedure or a function?

I believe I've seen both procedures and functions (elsewhere) named with Pascal case. I think I've only seen multi-word names on procedures, not functions.

I think including the `()` helps clarify that the thing being discussed is a function, as I don't think I've ever seen a `()` suffix on a procedure name.

All of which to say — I would prefer to keep the `()` and use Pascal- or camel-case to name functions, and to use multi-word names for procedures, and to always be explicit about `function` calls vs `procedure` runs.

msporny replied:
@Wind4Greg have @TallTed's issues been addressed? I'd like to merge this PR, but not until we've addressed @TallTed's concerns.

Wind4Greg replied:
Hi  @msporny not  quite sure what to do  about @TallTed's concerns some of  the  inconsistent naming comes  from the IETF  drafts. I'm not sure we differentiate  between  `functions` and `procedures` as terms.  If  there is  something  concrete  that needs to be changed  happy  to  update.

msporny replied:
We need to be consistent with what we call these things. As someone not familiar with some of this new text, I'd question what the difference is between a "function" and a "procedure". I'd be worried that I am missing something important when seeing those words if they're used interchangeably. IOW, we should pick one and be consistent in its usage. I suggest we use the word "algorithm" throughout all of the specifications as I believe that's what we do throughout the other specifications.

I try to avoid the use of the word "function" because some developers believe it to be referring to a software library call.

I avoid the word "procedure" because readers believe it's a loosely defined set of tasks where the implementation details are unimportant.

I use the word algorithm because it doesn't have the same connotations as "function" nor "procedure" -- it is more abstract than a function but less abstract than a procedure.

Can we update ALL of the references to refer to "algorithms" and NOT use the `()` notation because that implies a concrete function (which the IETF specifications are not).

TallTed commented on index.html:
```diff
@@ -1063,41 +1068,26 @@ <h4>serializeDerivedProofValue</h4>
                     <li>
 Initialize |proofValue| to start with the
-disclosure proof header bytes `0xd9`, `0x5d`, and `0x09`.
+disclosure proof header bytes `0xd9`, `0x5d`, and `0x07`.
```
```suggestion
disclosure proof header bytes: `0xd9`, `0x5d`, and `0x07`.
```

dlongley commented on index.html:
```diff
@@ -953,7 +958,7 @@ <h4>createDisclosureData</h4>
             <li>
 Return an object with properties matching |bbsProof|, "verifierLabelMap" for |labelMap|,
-|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, and |pseudonym|, if
+|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, |pseudonym|, and |L| if
```
We should use a better semantic name than `L` even if we want to say that it maps to `L` in a low-level BBS spec. The letter `L` itself can be confused with other symbols -- but at least this is less likely with an upper-case `L`.

Wind4Greg replied:
Hi @dlongley, *L* is the number of issuer signed BBS messages. Or in W3C VC-DI-BBS speak the number of non-mandatory n-quads. When working with Blind BBS and Pseudonym BBS *proofs* (W3C derived proofs) this is a new parameter that is needed during verification since we separated out the "group generators" for the blinded (holder) messages from the issuer messages to simplify indexing and to unify pseudonym proof verification. I'll take any suggestions for naming!

TallTed commented on index.html:
```diff
@@ -1556,18 +1543,19 @@ <h4>Base Proof Serialization (bbs-2023)</h4>
 cryptographically random 32 byte |pid| value. Compute the
-`bbsSignature` using the `Sign` procedure of [[CFRG-Pseudonym-BBS-Signature]],
-with appropriate key material, `bbsHeader` for the `header`, `bbsMessages`
+`bbsSignature` using the `Signer Provided PID Signature Generation`  procedure
```
```suggestion
`bbsSignature` using the `Signer Provided PID Signature Generation` procedure
```

dlongley commented on index.html:
```diff
@@ -1063,41 +1068,26 @@ <h4>serializeDerivedProofValue</h4>
-Initialize |components| to an array with elements containing the values of
 |bbsProof|, |compressedLabelMap|, |mandatoryIndexes|, |selectiveIndexes|,
-|presentationHeader|, and |pseudonym|.
+|presentationHeader| and |L|.
```
I see this changed from `pseudonym` to `L` ... I'm not checking this closely right now, but will do so during implementation.

Wind4Greg replied:
@dlongley both *anonymous holder binding* and *pseudonyms* features need the *L* parameter now.  Pseudonyms still need *pseudonyms* ;-).

dlongley reviewed (approved):
Approving, but see comments. Thanks!

TallTed reviewed (changes requested):
Just a few nits

Wind4Greg commented:
@TallTed (and @dlongley ) I updated to replace the *L* variable name with *lengthBBSMessages* and cleaned up its usage in the procedures related to verification. Note that the *Anonymous Holder Binding* feature needs the *lengthBBSMessages* (*L*) parameter, while both pseudonym features need the *pseudonym* **and** *lengthBBSMessages* (*L*) parameters. Hopefully I've gotten the conjunctions right for this.

TallTed commented on index.html:
```diff
@@ -925,7 +925,7 @@ <h4>createDisclosureData</h4>
 If |featureOption| equals `"anonymous_holder_binding"`,
-`"pseudonym_issuer_pid"`, or `"pseudonym_hidden_pid"` set the |L| parameter
+`"pseudonym_issuer_pid"`, or `"pseudonym_hidden_pid"`, set the |lengthBBSMessages| parameter
 equal to the length of the |bbsMessages| array.
```
```suggestion
to the length of the |bbsMessages| array.
```

TallTed commented on index.html:
```diff
@@ -958,7 +958,7 @@ <h4>createDisclosureData</h4>
             <li>
 Return an object with properties matching |bbsProof|, "verifierLabelMap" for |labelMap|,
-|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, |pseudonym|, and |L| if
+|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, |pseudonym|, and |lengthBBSMessages| if
 computed.
```
```suggestion
|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, |pseudonym|, and, if
computed, |lengthBBSMessages|.
```

TallTed commented on index.html:
```diff
@@ -1032,7 +1032,7 @@ <h4>serializeDerivedProofValue</h4>
 selective indexes (|selectiveIndexes|), and a BBS presentation header
 (|presentationHeader|), the |featureOption| indicator, and possibly a
-|pseudonym| and |L| value depending on the |featureOption| value.
+|pseudonym| and/or |lengthBBSMessages| value depending on the |featureOption| value.
```
Should probably include something about what `|featureOption| value` means there will be `a |pseudonym|`, `a |lengthBBSMessages|`, or both.
```suggestion
depending on the |featureOption| value, a |pseudonym| and/or |lengthBBSMessages| value.
```

TallTed commented on index.html:
```diff
@@ -1032,7 +1032,7 @@ <h4>serializeDerivedProofValue</h4>
 array of mandatory indexes (|mandatoryIndexes|), an array of
 selective indexes (|selectiveIndexes|), and a BBS presentation header
 (|presentationHeader|), the |featureOption| indicator, and possibly a
```
```suggestion
(|presentationHeader|), the |featureOption| indicator, and,
```

TallTed commented on index.html:
```diff
@@ -1111,8 +1111,8 @@ <h4>parseDerivedProofValue</h4>
 "labelMap", "mandatoryIndexes", "selectiveIndexes", "presentationHeader",
-"featureOption", and possibly "pseudonym" depending on the value of the
-|featureOption| parameter.
+"featureOption", and possibly "pseudonym" and/or "lengthBBSMessages" depending
+on the value of the |featureOption| parameter.
```
As above — what `value of the |featureOption| parameter` means there will be `a "pseudonym"`, `a "lengthBBSMessages"`, or both.
```suggestion
"featureOption", and, depending on the value of the |featureOption| parameter, 
"pseudonym" and/or "lengthBBSMessages".
```

TallTed commented on index.html:
```diff
@@ -1151,7 +1151,7 @@ <h4>parseDerivedProofValue</h4>
 Initialize `components` to an array that is the result of CBOR-decoding the
 bytes that follow the three-byte BBS disclosure proof header. If the result
-is not an array of five or six elements —
+is not an array of five to seven elements —
```
```suggestion
is not an array of five, six, or seven elements —
```

TallTed commented on index.html:
```diff
@@ -1165,7 +1165,8 @@ <h4>parseDerivedProofValue</h4>
             <li>
 Return <em>derived proof value</em> as an object with properties set to the five
 elements, using the names "bbsProof", "labelMap", "mandatoryIndexes",
```
```suggestion
Return <em>derived proof value</em> as an object with properties set to the five,
six, or seven elements, using the names "bbsProof", "labelMap", "mandatoryIndexes",
```

TallTed commented on index.html:
```diff
@@ -1238,7 +1240,7 @@ <h4>createVerifyData</h4>
 Return an object with properties matching |baseSignature|, |proofHash|,
 |nonMandatory|, |mandatoryHash|, |selectiveIndexes|, |featureOption|, and,
-possibly |pseudonym|.
+possibly |pseudonym| and/or |lengthBBSMessages|.
```
```suggestion
possibly, |pseudonym| and/or |lengthBBSMessages|.
```

TallTed commented on index.html:
```diff
@@ -1715,14 +1717,15 @@ <h4>Verify Derived Proof (bbs-2023)</h4>
 If the |featureOption| equals `"anonymous_holder_binding"`,
 initialize |verified| to the result of
 applying the verification algorithm `ProofVerify` algorithm of
```
```suggestion
applying the `ProofVerify` verification algorithm of
```

TallTed commented on index.html:
```diff
@@ -1715,14 +1717,15 @@ <h4>Verify Derived Proof (bbs-2023)</h4>
               <li>
 If the |featureOption| equals `"pseudonym"`, initialize |verified|
 to the result of
 applying the verification algorithm `ProofVerifyWithPseudonym` of
```
```suggestion
applying the `ProofVerifyWithPseudonym` verification algorithm of
```

TallTed reviewed (changes requested):
Some fixes to the changes, and some new changes. Grammar, punctuation, etc.

TallTed commented on index.html:
```diff
@@ -924,9 +924,13 @@ <h4>createDisclosureData</h4>
 to the length of the |bbsMessages| array.
+If |featureOption| equals `"pseudonym_issuer_pid"` set the |lengthBBSMessages|
+parameter to the length of the |bbsMessages| array + 1. <span class="note">This
+is due to the PID in the issuer known PID case being considered a "BBS message".
```
```suggestion
is due to the known-issuer PID being considered a "BBS message".
```

Wind4Greg commented:
Hi @dlongley and  @TallTed, I added  the updates  test vectors to the optional features to  this PR since it hasn't been merged yet  and  it would have been  confusing  to attach  them  elsewhere. Hence  to keep our procedures  sound  I'm asking for a re-review. **No** text has been changed  only the JSON test  vector  files  that are imported via ReSpec.

dlongley reviewed (approved):
Approving the additions with the same caveat that I haven't fully updated DB's implementation yet to run against them; I will do so in the near future and hopefully confirm interoperability. I bumped another comment around consistency in how we reference core BBS primitives that we will want to address at some point.

msporny commented:
Normative, multiple reviews, changes requested and made, no objections, merging.

TallTed commented on index.html:
```diff
@@ -923,6 +923,15 @@ <h4>createDisclosureData</h4>
+If |featureOption| equals `"anonymous_holder_binding"` or
+`"pseudonym_hidden_pid"`, set the |lengthBBSMessages| parameter
+to the length of the |bbsMessages| array.
+If |featureOption| equals `"pseudonym_issuer_pid"` set the |lengthBBSMessages|
```
```suggestion
If |featureOption| equals `"pseudonym_issuer_pid"`, set the |lengthBBSMessages|
```

TallTed replied:
overtaken

TallTed commented on index.html:
```diff
@@ -953,8 +962,8 @@ <h4>createDisclosureData</h4>
 Return an object with properties matching |bbsProof|, "verifierLabelMap" for |labelMap|,
-|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, and |pseudonym|, if
-computed.
+|mandatoryIndexes|, |selectiveIndexes|, |revealDocument|, |pseudonym|, and, if
+computed, |lengthBBSMessages|.
```
Is `|pseudonym|` now always computed?, or should rephrasing apply the `if computed` to both `|pseudonym|` and `|lengthBBSMessages|`?
